phoenix-core/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java

```diff
@@ -815,6 +818,10 @@ private void getCurrentRowStates(ObserverContext<RegionCoprocessorEnvironment> c
                                      BatchMutateContext context) throws IOException {
         Set<KeyRange> keys = new HashSet<KeyRange>(context.rowsToLock.size());
         context.dataRowStates = new HashMap<ImmutableBytesPtr, Pair<Put, Put>>(context.rowsToLock.size());
+        if (context.hasUncoveredIndex && !context.hasGlobalIndex && !context.hasTransform
```

Comment:
@kadirozde If there are multiple columns part of an uncovered index row key and we get a partial update which includes only a subset of those columns then in this case we still need to read the data table row to correctly generate the index row key.

Reply:
@tkhurana, You are right. I made the necessary change and included a test case for it. Currently, if a mutation in a mini batch is a partial update, the data table rows will read for all mutations in the batch. We can further optimize this in future by just reading rows only for partial mutations instead of all mutations.